### ISSUE TITLE
Rules run despite 'default = false' in configuration during tests

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#180]
## Description

When running integration tests with a configuration that sets `default = false` and explicitly enables only specific rules, some rules still run even though they should be disabled. This causes test failures and makes it difficult to test rules in isolation.

## Current Behavior

When using a configuration like:
```toml
[rules]
default = false

[rules.enabled]
MD024 = true

[MD024]
siblings_only = false
```

The following unexpected behavior occurs:
- MD024 runs as expected (enabled)
- MD025 runs unexpectedly (should be disabled)
- MDBOOK004 runs unexpectedly (should be disabled)

## Expected Behavior

Only explicitly enabled rules should run when `default = false` is set.

## Impact

This issue affects:
- Integration test reliability across all configuration batches
- Ability to test rules in isolation
- User experience when trying to enable only specific rules

## Reproduction Steps

1. Create a test markdown file with content that would trigger multiple rules
2. Create a configuration file with `default = false` and enable only one rule
3. Run mdbook-lint with the configuration
4. Observe that disabled rules still report violations

## Example

```bash
cd /tmp
cat > test.md << 'EOF'
# Heading

## Subheading

## Subheading

# Another Heading
EOF

cat > .mdbook-lint.toml << 'EOF'
[rules]
default = false

[rules.enabled]
MD024 = true
EOF

mdbook-lint lint -c .mdbook-lint.toml test.md
```

Result: MD025 and MDBOOK004 violations are reported even though only MD024 is enabled.

## Affected PRs

This issue affects test reliability in:
- #176 (Batch 1 configuration)
- #177 (Batch 2 configuration)
- #178 (Batch 3 configuration)
- #179 (Batch 4 configuration)

## Suggested Investigation

1. Check how `default = false` is processed in the configuration loading
2. Verify rule registry initialization when default is false
3. Check if certain rules are hardcoded to always run
4. Review the interaction between StandardRuleProvider and MdbookRuleProvider

## Priority

Medium - The configuration functionality works correctly in production, but this issue makes testing difficult and may confuse users who expect complete control over which rules run.
